### ISSUE TITLE
multiprocessing

### ISSUE THREAD

**Issue by davidwhogg** [aimalz/prob-z#4]
if you turn on multiprocessing in emcee, how much does it speed up your code?

**Comment by aimalz** [comment]
I finally managed to run my code on multiple cores, plotting as I go.  As expected, it takes less time, by an amount scaled approximately linearly by the number of processors used.  The big deal, though, is that the plots are made as I go, so I don't have to wait (much) for those at the end.  It took a long time for me to implement because I ran into a (known) bug between matplotlib, multiprocessing, and the way graphics are handles in OSX.

**Comment by davidwhogg** [comment]
You can fix this problem (in principle) by writing out output files and making your plots from the output files in a separate, independent process (even one that sits and waits for output data files!).

**Comment by aimalz** [comment]
Indeed, but the slow step isn't the computation.  Rather, saving the plots with matplotlib is the time-consuming step; it wasn't until I implemented multiprocessing that I discovered that the plotting itself (from saved files) isn't the problem.  (Edit: for example, in the latest run all the computation takes 7.5 minutes, all the plotting takes 9 seconds, and saving the plots takes 15 minutes.)  I'm looking into other plotting packages, including non-python ones, and will check in about that soonish.

**Comment by davidwhogg** [comment]
You can always fork the plotting -- we don't care about 15 minutes really.  We care about the hours and hours in the data center when we scale up.

**Comment by davidwhogg** [comment]
does it improve things to plot with "rasterized=True"?

**Comment by davidwhogg** [comment]
oh and if the problem is that matplotib is slow at saving _when it is in multiprocessing_ then that can be solved trivially:  By having the plotting code be in a separate python script that runs without multiprocessing.

**Comment by aimalz** [comment]
I already forked plotting and enabled rasterization (which made surprisingly little difference).  The times don't change when I just make the plots outside of multiprocessing from files created by emcee.  My point was more that saving the plots in general with matplotlib takes twice as long as doing the computation, which will presumably become problematic when the computation takes much longer.

**Comment by davidwhogg** [comment]
Can you point me to a very short plotting script and set of data that takes 15 minutes to save?  This seems long to me.  I know matplotlib is slow, but I have never seen times that long.  And are _all_ plots taking that long or just some?  The plots I have seen only have hundreds of line segments on them (or thousands) and such plots have never taken 15 minutes to save for me.  Maybe there is some particular plot that is taking longer than the others?

**Comment by aimalz** [comment]
Yes, the slow plots are the two of posterior samples, one being a fair bit slower than the other, because they have so many line segments.  I've got it at the limit of what I think makes for an interpretable plot, but perhaps others don't need as many samples to get the idea.  Output of one run (and the code that generates it) can be found [https://github.com/aimalz/prob-z/tree/master/nz/code](here); the slow plots take on the order of one minute per 1000 iterations of emcee with heavy thinning of the chain.

**Comment by davidwhogg** [comment]
This is posterior samples in a histogram or vs step number?  You can just drop this plot if you are computing the autocorrelation times for all the parameters.
